### PR TITLE
New tests for improved functions

### PR DESCRIPTION
**Pull request by siman**

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -15,45 +15,104 @@
...
+const ERROR_UNKNOWN_ENTITY_PROP_ID: &str = "Some of the provided property ids cannot be found on the current list of propery values of this entity";
+const ERROR_NO_ENTITY_PROP_IDS_ON_REMOVE: &str = "Cannot remove entity properties: an empty list of property ids provided";
+const ERROR_TEXT_PROP_IS_TOO_LONG: &str = "Text propery is too long";
+const ERROR_VEC_PROP_IS_TOO_LONG: &str = "Vector propery is too long";
+const ERROR_INTERNAL_RPOP_DOES_NOT_MATCH_ITS_CLASS: &str = "Internal property does not match its class";
+
+/// Length constraint for input validation
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize, Debug))]
+#[derive(Encode, Decode, Default, Clone, PartialEq, Eq)]
+pub struct InputValidationLengthConstraint {
```
Comment (bedeho) [inline src/lib.rs:66]: I thought you were use this from here

https://github.com/Joystream/substrate-utils

It would be a shame to inline this every time.

I am seeing that this is not actually yet in that library, so we could just forget it for now. Then go back and fix later.

File: src/lib.rs [hunk 2]
```diff
@@ -15,45 +15,104 @@
...
-//     block: BlockNumber,
-//     time: Moment
-// }
-
 #[cfg_attr(feature = "std", derive(Serialize, Deserialize, Debug))]
 #[derive(Encode, Decode, Default, Clone, PartialEq, Eq)]
 pub struct Class {
     id: ClassId,
 
+    /// All properties that has been used on this class across different class schemas.
```
Comment (bedeho) [inline src/lib.rs:113]: has => have

File: src/lib.rs [hunk 3]
```diff
@@ -168,13 +292,20 @@ decl_event!(
...
         <T as system::Trait>::AccountId,
     {
-        ClassCreated(AccountId, ClassId),
-        ClassPropertyAdded(AccountId, ClassId),
-        ClassSchemaAdded(AccountId, ClassId),
-
-        EntityCreated(AccountId, EntityId),
-        EntityUpdated(AccountId, EntityId),
-        EntityDeleted(AccountId, EntityId),
+        ClassCreated(ClassId),
```
Comment (bedeho) [inline src/lib.rs:295]: I am not 100% sure, but I believe I did point out that this module should not have events in the prior review. We have tried to follow the rule that non-public modules, i.e. modules without extrinsics, have no events. The caller has to layer them on top in a way which is use case specific.

I think that its fine to just keep this here for now though, given that it all works, and it would just be make-work to move it into the permissions module. We can revise in the future.

File: src/lib.rs [hunk 4]
```diff
@@ -105,16 +164,40 @@ pub struct Property {
...
     Bool,
     Uint16,
     Uint32,
     Uint64,
     Int16,
     Int32,
     Int64,
-    Text(u32),
+    Text(u16),
     Internal(ClassId),
```
Comment (bedeho) [inline src/lib.rs:177]: I added writeup of future work on versioned references here https://github.com/Joystream/substrate-versioned-store/issues/14 ,  but not need for that now.

File: src/lib.rs [hunk 5]
```diff
@@ -184,52 +315,567 @@ decl_module! {
...
+        Self::ensure_known_entity_id(entity_id)?;
+
+        let (entity, class) = Self::get_entity_and_class(entity_id);
+
+        // Check that schema id is not yet added to this entity:
+        let schema_not_added = entity.in_class_schema_indexes.iter().position(|x| *x == schema_id).is_none();
+        ensure!(schema_not_added, ERROR_SCHEMA_ALREADY_ADDED_TO_ENTITY);
+
+        // Check that schema_id is a valid index of class schemas vector:
+        let known_schema_id = schema_id < class.schemas.len() as u16;
```
Comment (bedeho) [inline src/lib.rs:478]: This guard should probably be before the prior one on 473. Gives the same result, but makes more sense that way, and avoids some runtime needless iteration if the schema is just invalid.

File: src/lib.rs [hunk 6]
```diff
@@ -184,52 +315,567 @@ decl_module! {
...
-        // Self::deposit_event(RawEvent::ClassCreated(account.clone(), class_id));
+        <EntityById<T>>::mutate(entity_id, |entity| {
+            entity.name = new_name;
+        });
+
+        Self::deposit_event(RawEvent::EntityNameUpdated(entity_id));
+        Ok(())
+    }
+
+    pub fn add_schema_support_to_entity(
```
Comment (bedeho) [inline src/lib.rs:463]: This should not allow the caller to _overwrite_ property values which already exist. Adding support for a schema should only involve providing initial values for any properties values which are being introduced for the entity for the first time.

This is not only most precisely mirrors the actual meaning of the extrinsic name, but more importantly, the permission system only understands the concept of  "adding schema support", not whether old property values are also being changed. This means that the permissions cannot guard against a user abusively editing old values that they should not be able to edit.

Lastly, it is also leading to a lot of complexity in the routine.

File: src/lib.rs [hunk 7]
```diff
@@ -184,52 +315,567 @@ decl_module! {
...
+        <EntityById<T>>::mutate(entity_id, |entity| {
+            entity.in_class_schema_indexes.push(schema_id);
+            entity.values = updated_values_with_nones;
+        });
+
+        Self::deposit_event(RawEvent::EntitySchemaAdded(entity_id));
+        Ok(())
+    }
+
+    pub fn update_entity_properties(
```
Comment (bedeho) [inline src/lib.rs:558]: Would the name `update_entity_property_values` be even more precise perhaps? The properties live on the class level after all.

File: src/lib.rs [hunk 8]
```diff
@@ -184,52 +315,567 @@ decl_module! {
...
+        entity_id: EntityId,
+        new_property_values: Vec<ClassPropertyValue>
+    ) -> dispatch::Result {
+
+        Self::ensure_known_entity_id(entity_id)?;
+
+        let (entity, class) = Self::get_entity_and_class(entity_id);
+
+        ensure!(!entity.in_class_schema_indexes.is_empty(), ERROR_ENTITY_DOES_NOT_SUPPORT_SCHEMAS_YET);
```
Comment (bedeho) [inline src/lib.rs:567]: This explicit test does not make sense to do explicitly. The `new_property_values` could still reference schemas that don't or which have not been added support for, so explicit checks are always going to be needed a worst case on any input from an adversary. Non-adversary would not provide input that was blocked on this guard anyway.

File: src/lib.rs [hunk 9]
```diff
@@ -184,52 +315,567 @@ decl_module! {
...
+
+        let mut updated_values = entity.values;
+
+        for new_prop_value in new_property_values.iter() {
```
Comment (bedeho) [inline src/lib.rs:571]: This looks right, but a but more comments would be great.

File: src/lib.rs [hunk 10]
```diff
@@ -127,6 +210,8 @@ impl Default for PropertyType {
 #[derive(Encode, Decode, Clone, PartialEq, Eq)]
 pub enum PropertyValue {
     None,
```
Comment (bedeho) [inline src/lib.rs:212]: Why is `None` an option here? If its an artifact of the need for default constructability then remove that, we should not have default constructibility either.

Comment (bedeho) [reply src/lib.rs:212]: One can make a similar but weaker argument about `PropertyValue::None`. None is not an actual value for any of the types. None is only a possible value in the context of satisfying a schema or not. The direct way to model that would be in `ClassPropertyValue`, which models this directly, so it would be

 ```Rust
 struct ClassPropertyValue {
value: Option<PropertyValue>
}
```
but I don't know if it worth changing this now.

File: src/lib.rs [hunk 11]
```diff
@@ -184,52 +315,567 @@ decl_module! {
...
+            entity.values = updated_values;
+        });
+
+        Self::deposit_event(RawEvent::EntityPropertiesUpdated(entity_id));
+        Ok(())
+    }
+    
+    /// Only non required property values can be removed.
+    /// In fact when removing a property value, it is replaced with PropertyValue::None.
+    pub fn remove_entity_properties(
```
Comment (bedeho) [inline src/lib.rs:601]: First, given that this does in fact not remove properties from entities, but instead sets property values to be none/not-set, calling it `remove_entity_properties` is not a great name.

Second, why is this needed, can't one just use `update_entity_properties` for this?

**Review by bedeho — CHANGES_REQUESTED** [review]

File: src/lib.rs (hunk 9 above)
Comment (siman) [reply src/lib.rs:571]: Added more comments here: https://github.com/siman/substrate-versioned-store/commit/c297f9fd73fa66321740d8dc0579c9198121b26a

File: src/lib.rs (hunk 5 above)
Comment (siman) [reply src/lib.rs:478]: Refactored here https://github.com/siman/substrate-versioned-store/commit/0e44b7bfdb990f95b4443b3f5647115afb978333

File: src/lib.rs (hunk 6 above)
Comment (siman) [reply src/lib.rs:463]: Resolved here https://github.com/Joystream/substrate-versioned-store-module/pull/16/commits/6bac3c1006457a9494313b0515b481b6d9ac9522

File: src/lib.rs (hunk 7 above)
Comment (siman) [reply src/lib.rs:558]: Renamed here: https://github.com/Joystream/substrate-versioned-store-module/pull/16/commits/0328d190df1a0438cc0589546a3ddff6b377a564

File: src/lib.rs (hunk 10 above)
Comment (siman) [reply src/lib.rs:212]: Substrate requires a default value for enums that should be encoded/decoded (for example used on UI). That's why I need to add a default value for both enums: `PropertyType` and `PropertyValue`

File: src/lib.rs (hunk 11 above)
Comment (siman) [reply src/lib.rs:601]: Function `remove_entity_properties` deleted here https://github.com/Joystream/substrate-versioned-store-module/pull/16/commits/e6e62165e1ae5e64123f485f5e2d4141d020a5c2

File: src/lib.rs (hunk 8 above)
Comment (siman) [reply src/lib.rs:567]: This check for supported schemas is removed with this commit https://github.com/Joystream/substrate-versioned-store-module/pull/16/commits/fbad5cf5c6c9f965bc8cd4f29428de01a7872076

File: src/lib.rs (hunk 1 above)
Comment (siman) [reply src/lib.rs:66]: Created a corresponding issue: 
https://github.com/Joystream/substrate-versioned-store-module/issues/17

**Comment by siman** [comment]
@bedeho a new PR with all fixes is here:
https://github.com/Joystream/substrate-versioned-store-module/pull/16

File: src/lib.rs (hunk 10 above)
Comment (bedeho) [reply src/lib.rs:212]: Ok, lets just stick with this for now.
